hack/gmpctl/lib.sh

```diff
@@ -0,0 +1,665 @@
+# limitations under the License.
+
+# NOTE for contributors: Bash is funky, but sometimes more readable than Go/easier to iterate.
+# Eventually, we could rewrite more critical pieces to Go, but you're welcome to add some quick
+# pieces in bash to automate some stuff.
+#
+# See README.md#bash for rules to writing bash.
+
+set -o errexit
+set -o pipefail
+set -o nounset
+
+if [[ -n "${DEBUG_MODE:-}" ]]; then
+	set -o xtrace
+fi
+
+log_err() {
+	echo "❌  ${1}" >&2
+}
+
+# TODO(bwplotka): Finding correct script dir is not so trivial.
+if [[ -z "${SCRIPT_DIR}" ]]; then
+	log_err "SCRIPT_DIR envvar is required."
+	return 1
+fi
+
+release-lib::confirm() {
+	local prompt_message="${1:-Are you sure?}"
+
+	# -p: Display the prompt string.
+	# -r: Prevents backslash interpretation.
+	# -n 1: Read only one character.
+	read -p "$prompt_message [y/n/CTR+C]: " -r -n 1 response
+	echo # Ensures the cursor moves to the next line after input.
+	case "$response" in
+	[yY])
+		return 0
+		;;
+	[nN])
+		log_err "The action has been cancelled as requested."
+		return 1
+		;;
+	*)
+		log_err "Invalid input. Exiting script."
+		exit 1
+		;;
+	esac
+}
+
+release-lib::idemp::vulnlist() {
+	local dir="${1}"
+	if [[ -z "${dir}" ]]; then
+		log_err "dir arg is required."
+		return 1
+	fi
+	local vuln_file="${2}"
+	if [[ -z "${vuln_file}" ]]; then
+		log_err "vuln_file arg is required."
+		return 1
+	fi
+	if [[ "${vuln_file}" != /* ]]; then
+		log_err "vuln_file arg must point to an absolute file path."
+		return 1
+	fi
+
+	# TODO(bwplotka): We could ask user if we should reuse existing vulnfile.
+	release-lib::vulnlist "${dir}" "${vuln_file}"
+}
+
+release-lib::vulnlist() {
+	local dir="${1}"
+	if [[ -z "${dir}" ]]; then
+		log_err "dir arg is required."
+		return 1
+	fi
+	local vuln_file="${2}"
+	if [[ -z "${vuln_file}" ]]; then
+		log_err "vuln_file arg is required."
+		return 1
+	fi
+	if [[ "${vuln_file}" != /* ]]; then
+		log_err "vuln_file arg must point to an absolute file path."
+		return 1
+	fi
+
+	readarray -t DOCKERFILES < <(release-lib::dockerfiles "${dir}")
+	local go_version=$(release-lib::dockerfile_go_version "${DOCKERFILES[0]}")
+	if [[ -z "${go_version}" ]]; then
+		log_err "can't find any golang image in ${DOCKERFILES[0]}"
+		return 1
+	fi
+
+	echo "🔄  Detecting Go ${go_version} vulnerabilities to fix..."
+	pushd "${SCRIPT_DIR}/vulnupdatelist/"
+	if [[ ! -f "./api.text" ]]; then
+		log_err "$(pwd)/api.text file not found in your filesystem. Please create it with your NVD API key. See https://nvd.nist.gov/developers/request-an-api-key"
+		return 1
+	fi
+
+	go run "./..." \
+		-go-version=${go_version} \
+		-only-fixed \
+		-dir="${dir}" \
+		-nvd-api-key="$(cat "./api.text")" | tee "${vuln_file}"
+	if [[ -z $(cat "${vuln_file}") ]]; then
+		# Print this, otherwise error on the above might keep this file mistakenly empty.
+		echo "no vulnerabilities" >"${vuln_file}"
+	fi
+	popd
+}
+
+release-lib::gomod_vulnfix() {
```

Comment:
@hsmatulis feedback

Functions across bash files.

For Go -> bash files it might easier.

Reply:
From @hsmatulis:

* libs.sh 600 lines in a big file.
* More explicit entry point where it's entry point. (internal vs external functions)